### ISSUE TITLE
How do I make percentages work? I'm struggling insanely

### ISSUE THREAD

**Issue by Nazor1** [notKamui/Keval#51]
So I have a viewmodel for my calculator where I handle the calculator expressions. When user presses %, they see % but the calculator from my experience doesn't understand that, so I have internal expression which should translate it to *0.01.

The problem:
For example 50-10% does not work at all. It should be 45 on normal calculator but it results in 49.9 on mine for some reason. What's crazy is that only multiplication works, for example "2*4%=0.08" (which is correct). Other operators do not work " + , - , / ".

```kotlin
// Handles the percent operation
    private fun handlePercent() {
        val currentValue = currentExpression.value
        if (currentValue.isNotEmpty() && currentValue.last().isDigit()) {
            currentExpression.value += "%"
            internalExpression += "*0.01"
        }
    }

    // Preprocess percentages in the expression
    private fun preprocessPercentages(expression: String): String {
        val regex = Regex("(\\d+\\.?\\d*)%")
        return regex.replace(expression) { matchResult ->
            val percentageValue = matchResult.groupValues[1]

            // Find the preceding expression to use as the base for the percentage calculation
            val precedingExpression = expression.substring(0, matchResult.range.first).trimEnd()
            val lastOperatorIndex = precedingExpression.lastIndexOfAny(charArrayOf('+', '-', '*', '/'))

            // Extract base number and correct the expression
            val baseExpression = if (lastOperatorIndex != -1) {
                precedingExpression.substring(0, lastOperatorIndex).trim()
            } else {
                "1" // Default to multiplying by 1 if no operator is found
            }

            // Replace with the correct percentage calculation
            "($baseExpression * $percentageValue * 0.01)"
        }
    }

    private fun evaluateExpression() {
        if (!canEvaluateExpression()) return

        try {
            // Preprocess the expression to handle percentages correctly
            val processedExpression = preprocessPercentages(internalExpression)

            // Evaluate the processed expression
            val result = Keval.eval(processedExpression)

            // Convert the result to BigDecimal for precise rounding
            val bigDecimalResult = BigDecimal(result).setScale(9, RoundingMode.HALF_UP).stripTrailingZeros()

            // Convert to string, handling cases like "336.0" -> "336"
            resultExpression.value = if (bigDecimalResult.scale() <= 0) {
                bigDecimalResult.toPlainString()
            } else {
                bigDecimalResult.toString()
            }

            // Update currentExpression and internalExpression with the result
            currentExpression.value = resultExpression.value
            internalExpression = resultExpression.value

        } catch (e: KevalZeroDivisionException) {
            resultExpression.value = "Error: You can't divide by zero"
        } catch (e: KevalInvalidArgumentException) {
            resultExpression.value = "Error: Invalid argument"
        } catch (e: KevalInvalidExpressionException) {
            resultExpression.value = "Error: Invalid expression"
        } catch (e: KevalInvalidSymbolException) {
            resultExpression.value = "Error: Invalid symbol '${e.invalidSymbol}'"
        } catch (e: KevalDSLException) {
            resultExpression.value = "Error: Invalid DSL configuration"
        } catch (e: Exception) {
            resultExpression.value = "Error: Invalid calculation"
        }
    }
}
```

Result expression is what shows as result in little text like Samsung or Google calculator. Current expression is self-explanatory. Internal expression is what calculator translates to itself, for example the division symbol or percentage.

I'm dying at this point

**Comment by notKamui** [comment]
It's quite simple actually

![image](https://github.com/user-attachments/assets/1a057ada-ec4e-4afc-9fdd-8e9b05469787)

the `%` is not a percentage unary operator, but the "remainder/modulo" operator, like in most programming languages, including Kotlin, just like "multiplication" is `*` and not `×`.

`%` is a binary operator which will return the remainder of the left operand divided by the right operand (again, like in most programming languages)

If you want `%` to be a "percentage unary operator", you need to provide your own implementation. Please refer to the section about customization in the README.

Here is an example:

```kt
val KEVAL_INSTANCE = Keval.create { // builder instance
    includeDefault() // this function includes the built-in resources
    
    unaryOperator { // this function adds a unary operator ; you can call it several times
        symbol = '%'
        isPrefix = false // make it postfix
        implementation = { it / 100 }
    }
}

KEVAL_INSTANCE.eval("2 * 4%") // 0.08
```

In fact, I recommend you don't use `includeDefault()` and implement all the operators you want yourself since you're making a calculator. The parser shouldn't bundle functions that will never be used.

---

As for the second issue, please refer to this issue comment: https://github.com/sosauce/CuteCalc/issues/64#issuecomment-1994596913

**Comment by notKamui** [comment]
TLDR: not a bug, but misusage

**Comment by notKamui** [comment]
By the way, if you do want contextual percentage (which again, in expression evaluators, is ambiguous and not normative), you could pretreat the string before passing it to Keval. But this is not in scope of Keval's goal, so it will never be a feature.

**Comment by notKamui** [comment]
```kt
val KEVAL_INSTANCE = Keval.create {
   ...
    
    function { // declares a custom function
        name = "percent"
        arity = 2
        implementation = { (x, percent) -> x * (1 + (percent / 100)) }
    }
}

KEVAL_INSTANCE.eval("percent(50, -10)") // 45
```

**Comment by notKamui** [comment]
![image](https://github.com/user-attachments/assets/da8c5a33-9f46-434a-bebf-22d17d33a09a)

**Comment by Nazor1** [comment]
Thank you so much. You're bringing my sanity down.

Well I successfully made it so basic expressions like 2/4% or 50-10% are giving me a correct answer. I have another problem which is that I have no idea how to make support for more complex percent expressions, like 5+5%+5% like you can do a lot of these percentages on samsung calculator

```kotlin
// Preprocess percentages in the expression
    private fun preprocessExpression(expression: String): String {
        val regex = Regex("(\\d+\\.?\\d*)([+\\-*/])(\\d+\\.?\\d*)%")
        return regex.replace(expression) { matchResult ->
            val baseValue = matchResult.groupValues[1]
            val operator = matchResult.groupValues[2]
            val percentageValue = matchResult.groupValues[3]

            // Convert to the percent function format
            when (operator) {
                "+" -> "percent($baseValue, $percentageValue)"
                "-" -> "percent($baseValue, -$percentageValue)"
                "*" -> "$baseValue * ($percentageValue / 100)"
                "/" -> "$baseValue / ($percentageValue / 100)"
                else -> matchResult.value // Leave unchanged if no operator matched
            }
        }
    }

    // Define a custom Keval instance with the custom percent function
    private val kevalInstance = Keval.create {
        includeDefault() // Include default operators and functions

        function { // Define the custom 'percent' function
            name = "percent"
            arity = 2
            implementation = { (x, percent) -> x * (1 + (percent / 100)) }
        }
    }

    private fun evaluateExpression() {
        if (!canEvaluateExpression()) return

        try {
            // Preprocess the expression to handle percentages correctly
            val processedExpression = preprocessExpression(internalExpression)

            // Use the custom Keval instance to evaluate the expression
            val result = kevalInstance.eval(processedExpression)

            // Convert the result to BigDecimal for precise rounding
            val bigDecimalResult = BigDecimal(result).setScale(9, RoundingMode.HALF_UP).stripTrailingZeros()

            // Convert to string, handling cases like "336.0" -> "336"
            resultExpression.value = if (bigDecimalResult.scale() <= 0) {
                bigDecimalResult.toPlainString()
            } else {
                bigDecimalResult.toString()
            }

            // Update currentExpression and internalExpression with the result
            currentExpression.value = resultExpression.value
            internalExpression = resultExpression.value

        } catch (e: KevalZeroDivisionException) {
            resultExpression.value = "Error: You can't divide by zero"
        } catch (e: KevalInvalidArgumentException) {
            resultExpression.value = "Error: Invalid argument"
        } catch (e: KevalInvalidExpressionException) {
            resultExpression.value = "Error: Invalid expression"
        } catch (e: KevalInvalidSymbolException) {
            resultExpression.value = "Error: Invalid symbol '${e.invalidSymbol}'"
        } catch (e: KevalDSLException) {
            resultExpression.value = "Error: Invalid DSL configuration"
        } catch (e: Exception) {
            resultExpression.value = "Error: Invalid calculation"
        }
    }
```

**Comment by notKamui** [comment]
this is an extremely complex issue.

your samsung calculator can do it easily because it is not an expression evaluator. Like most calculators, it sequential : you enter a value, then an operator, and another operand, and it outputs the result, and then you can continue again.

this is effectively equivalent to calculating things left to right, ignoring any precedence.

in such case, there are no ambiguities.

but this is not our case here.

the only idea I have is to pretreat the string to add parentheses to the expression, with regex lookbacks. this still seems extremely hard to me.

the other solution is to make a sequential calculator, and not an evaluator. at which point, Keval has no purpose.

**Comment by notKamui** [comment]
to put it into perspective, in the real world: if you look at scientific calculators, which are NOT sequential, but evaluators like Keval, they don't have the percentage operator, because of this issue with precedence.

**Comment by Nazor1** [comment]
Oooh... now I understand. Damn, do you know any sequential calculator libraries? I instantly loved Keval because of how simple & good it is, so maybe there's something also like this.
